theano/sandbox/gpuarray/basic_ops.py

```diff
@@ -599,50 +681,53 @@ def c_code(self, node, name, inputs, outputs, sub):
     def c_code_cache_version(self):
         return (3,)
 
-
 cuda_from_gpu = CudaFromGpu()
 
 
 class GpuAlloc(HideC, Alloc):
-    def __init__(self, memset_0=False):
+    __props__ = ('memset_0', 'context') + Alloc.__props__
+
+    context_type = gpu_context_type
+
+    def __init__(self, context, memset_0=False):
         """memset_0 is only an optimized version. True, it mean the
         value is always 0, so the c code call memset as it is faster.
         """
+        self.context = context
         self.memset_0 = memset_0
 
-    def __eq__(self, other):
-        return type(self) == type(other) and self.memset_0 == other.memset_0
-
-    def __hash__(self):
-        return hash(type(self)) ^ hash(self.memset_0)
-
     def __str__(self):
         #Hide the memset parameter when not used to prevent confusion.
         if self.memset_0:
-            s = "%s{memset_0=%s}" % (self.__class__.__name__, self.memset_0)
+            s = "%s<%s>{memset_0=%s}" % (self.__class__.__name__,
+                                         self.context, self.memset_0)
         else:
-            s = self.__class__.__name__
+            s = "%s<%s>" % (self.__class__.__name__, self.context)
         return s
 
     def make_node(self, value, *shape):
         res = Alloc.make_node(self, value, *shape)
-        value = as_gpuarray_variable(value)
+        value = as_gpuarray_variable(value, context=self.context)
         otype = GpuArrayType(dtype=res.outputs[0].dtype,
-                             broadcastable=res.outputs[0].broadcastable)
+                             broadcastable=res.outputs[0].broadcastable,
+                             context=self.context)
         return Apply(self, [value] + res.inputs[1:], [otype()])
 
+    def get_context(self, node):
+        return self.context
+
     def c_headers(self):
         return ['<numpy_compat.h>']
 
-    def perform(self, node, inputs, outs):
+    def perform(self, node, inputs, outs, ctx):
         out, = outs
         v = inputs[0]
         sh = tuple(map(int, inputs[1:]))
         if out[0] is None or out[0].shape != sh:
```

Comment:
I think you should check that the context of out is the right one. Just in case. We are starting to have more preallocated output and this could happen in case of errors.